### PR TITLE
Implement stepIn requests in debugger

### PR DESCRIPTION
**Pull request by JWZ2018**
Release note: none
Summary:
-implement step in requests in the debugger: set a flag when a stepIn request is received, on each evaluate call on an ast node, check whether that node is a statement and it can be stopped on
-remove a redundant PREPACK_FINISHED message from Prepack to the adapter that interferes with the stepIn request on the last statement. The adapter already tracks when Prepack is finished by listening for the "exit" event.

Test Plan:
```
function a(num) {
  if (num === 3) {
    var y = 5;
  } else {
    {
      let abc = 3;
    }
    let z = [1,2,3];
    let e = Symbol("xyz");
    let f = {
      g: 0,
      h: 1,
      i: {
        j: 2,
        k: 3,
      },
    };
    let x = 6;
  }
}
a();
```
Breakpoints on line 2 and stepIn requests until the end of the program.
Verified in both CLI and IDE.

### CODE REVIEW

File: src/methods/is.js [hunk 1]
```diff
@@ -350,3 +351,42 @@ export function IsStatic(classElement: BabelNodeClassMethod): boolean {
   // $FlowFixMe need to backport static property to BabelNodeClassMethod
   return classElement.static;
 }
+
+export function IsStatement(node: BabelNode): boolean {
+  if (node.type === "BlockStatement") return true;
```
Comment (hermanventer) [inline src/methods/is.js]: You can use a switch statement instead of this long list of ifs.

File: src/debugger/Debugger.js [hunk 2]
```diff
@@ -127,6 +131,37 @@ export class DebugServer {
     }
   }
 
+  checkStepIn(ast: BabelNode) {
+    if (this._steppingInfo !== undefined) {
+      if (this.isValidSteppingLocation(ast)) {
+        if (ast.loc && ast.loc.source) {
```
[the remark this answers is not in the document]
Comment (JWZ2018) [reply src/debugger/Debugger.js]: Flow will complain when I remove this check.

File: src/debugger/channel/DebugChannel.js [hunk 3]
```diff
@@ -89,7 +89,12 @@ export class DebugChannel {
     this.writeOut(this._marshaller.marshallVariablesResponse(requestID, variables));
   }
 
-  sendPrepackFinish(): void {
-    this.writeOut(this._marshaller.marshallPrepackFinish());
+  sendStepInResponse(filePath: string, line: number, column: number) {
```
Comment (yinghuitan) [inline src/debugger/channel/DebugChannel.js]: Instead of sendStepInResponse(), I suggest you create a single debuggerStop message and with a field containing the stop reason(breakpoint, or step-into, step-over, step-out etc...). This aligns with VSP StoppedEvent.

File: src/debugger/Debugger.js [hunk 4]
```diff
@@ -271,14 +311,25 @@ export class DebugServer {
     this._channel.sendVariablesResponse(requestID, variables);
   }
 
+  processStepInCommand(requestID: number, ast: BabelNode) {
```
Comment (yinghuitan) [inline src/debugger/Debugger.js]: You may want to generalize this into "_ processStepCommand" later.

File: src/debugger/Debugger.js [hunk 5]
```diff
@@ -127,6 +131,37 @@ export class DebugServer {
...
+        if (ast.loc && ast.loc.source) {
+          this._channel.sendStepInResponse(ast.loc.source, ast.loc.start.line, ast.loc.start.column);
+          this._steppingInfo = undefined;
+          this.waitForRun(ast);
+        }
+      }
+    }
+  }
+
+  isValidSteppingLocation(ast: BabelNode): boolean {
```
Comment (yinghuitan) [inline src/debugger/Debugger.js]: The name does not make sense. The meaning of this method should be "shouldCompleteStep()" or something similar.

Comment (yinghuitan) [inline src/debugger/Debugger.js]: Each stepping mode will have its own stop condition, so you may want to generalize each step mode into its own Stepper class.

File: src/debugger/Debugger.js [hunk 6]
```diff
@@ -127,6 +131,37 @@ export class DebugServer {
...
+    if (!IsStatement(ast)) return false;
```
Comment (yinghuitan) [inline src/debugger/Debugger.js]: Add comment why IsStatement() is needed here.

File: src/debugger/SteppingManager.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+      if (ast.loc && ast.loc.source) {
+        this._stepInData = undefined;
+        this._channel.sendPrepackStopped("Step In", ast.loc.source, ast.loc.start.line, ast.loc.start.column);
+        return true;
+      }
+    }
+    return false;
+  }
+
+  _isValidStepInLocation(ast: BabelNode): boolean {
```
Comment (yinghuitan) [inline src/debugger/SteppingManager.js]: As I said in previous comment, this is misleading name. Every location is a valid step-in location. This function checks if step-in request completes. So rename to "_isStepInComplete".

File: src/debugger/adapter/DebugAdapter.js [hunk 8]
```diff
@@ -247,6 +254,12 @@ class PrepackDebugSession extends LoggingDebugSession {
       }
     );
   }
+
+  stepInRequest(response: DebugProtocol.StepInResponse, args: DebugProtocol.StepInArguments): void {
```
Comment (yinghuitan) [inline src/debugger/adapter/DebugAdapter.js]: Add comment "// override" for all these methods so that reviewer can easily see how this method is used/called.

File: src/debugger/channel/MessageMarshaller.js [hunk 9]
```diff
@@ -294,24 +306,24 @@ export class MessageMarshaller {
...
+      invariant(result.hasOwnProperty("line"));
+      invariant(!isNaN(result.line));
+      invariant(result.hasOwnProperty("column"));
+      invariant(!isNaN(result.column));
+      let dbgResponse: DebuggerResponse = {
+        id: requestID,
+        result: result,
+      };
+      return dbgResponse;
+    } catch (e) {
```
Comment (yinghuitan) [inline src/debugger/channel/MessageMarshaller.js]: This is ugly. I suggest you take out the message parsing "JSON.parse()" into a method, which needs try catch. Then all _unmarshalXXX methods do not need try...catch.

File: src/debugger/Debugger.js [hunk 10]
```diff
@@ -71,8 +75,8 @@ export class DebugServer {
   // Checking if the debugger needs to take any action on reaching this ast node
   checkForActions(ast: BabelNode) {
     this.checkForBreakpoint(ast);
-
-    // last step: set the current location as the previously executed line
+    this.checkStepIn(ast);
```
Comment (yinghuitan) [inline src/debugger/Debugger.js]: It's fine right now, but you will soon change it to checkStepComplete() general method.

Comment (yinghuitan) [inline src/debugger/Debugger.js]: Request change: This kind of sequencial checking has a problem:
Assume you are at line 10, and you set a breakpoint at line 11. Then you issue a step-into which will land at line11 later. In current design, debugger will stop twice. Once inside checkForBreakpoint(), when user resumes from breakpoint, checkStepIn() will stop it again which is wrong.

Comment (JWZ2018) [reply src/debugger/Debugger.js]: This problem should happen in the current code. But strangely, I'm not able to reproduce it. Will look into it more.

Comment (JWZ2018) [reply src/debugger/Debugger.js]: There is a check in _isValidSteppingLocation for whether the current location is the same as the previously stopped location. If so, it will not stop again.

File: src/debugger/channel/MessageMarshaller.js (hunk 9 above)
Comment (JWZ2018) [reply src/debugger/channel/MessageMarshaller.js]: I will do this and the refactoring of marshallXXX in separate PRs to not make this one too big.

File: src/debugger/SteppingManager.js [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+  processStepCommand(kind: "in" | "over" | "out", currentNode: BabelNode) {
+    if (kind === "in") {
+      this._processStepIn(currentNode);
+    }
+    // TODO: implement stepOver and stepOut
+  }
+
+  _processStepIn(ast: BabelNode) {
+    invariant(ast.loc && ast.loc.source);
```
Comment (yinghuitan) [inline src/debugger/SteppingManager.js]: Add an invariant here: invariant(this. _stepInData === undefined);

File: src/debugger/Debugger.js (hunk 10 above)
Comment (yinghuitan) [reply src/debugger/Debugger.js]: It works in this scenario because you are missing one invariant here: checking "this. _stepInData is undefined" in _processStepIn() which means there should only be one step-in request on the flight. Currently, you are silently overwriting the this._stepInData for each step-in instead of explicitly cancel the previous stepper. After adding this invariant, you will see the failure. 
Another way to demonstrate the failure:
1. Current line is 10, while set breakpoint at line 11
2. Step-in from line10 => line11
3. BreakpointManager stops at line11 while step-into stepper is not done yet.
4. You do a "step-over"(note: note step-into) which will not overwrite this._stepInData. 
5. The code comes back and checkStepInto() which will stop again at line11.

The root cause of the problem is that it sequentially notify each managers. Instead, you want all the execution control related managers to be notified and then consolidate a final stop disposition decision.

In summary, you want:
1. Separate stepping logic into its own stepper.
2. At each line, send the current AST to all managers
3. Collect stop decision from each manager and join them into a single decision.

File: src/debugger/channel/MessageMarshaller.js [hunk 12]
```diff
@@ -43,13 +44,15 @@ export class MessageMarshaller {
...
   }
 
-  marshallBreakpointStopped(args: Breakpoint): string {
-    return `${this
-      ._lastRunRequestID} ${DebugMessage.BREAKPOINT_STOPPED_RESPONSE} ${args.filePath} ${args.line} ${args.column}`;
-  }
-
-  marshallPrepackFinish(): string {
-    return `${this._lastRunRequestID} ${DebugMessage.PREPACK_FINISH_RESPONSE}`;
+  marshallPrepackStopped(reason: PrepackStoppedReason, filePath: string, line: number, column: number): string {
```
Comment (yinghuitan) [inline src/debugger/channel/MessageMarshaller.js]: It's weird to have "Prepack" in this. Let's call it, marshallStoppedResponse or marshallStoppedEvent

**Comment by JWZ2018** [comment]
I reproduced the bug with stopping multiple times. It only occurs when I step to a line with a breakpoint and then try to continue from there. I will fix it.

**Comment by JWZ2018** [comment]
I've fixed the buggy behaviour of stopping multiple times at the same location.

**Comment by JWZ2018** [comment]
Only the step in is cancelled during a stop currently. We can handle other kinds of stepping differently on a stop.
